core/lib/dal/src/connection/holder.rs

```diff
@@ -2,27 +2,96 @@
 use std::fmt;
 // External imports
 use sqlx::pool::PoolConnection;
-use sqlx::{postgres::Postgres, PgConnection, Transaction};
+use sqlx::{postgres::Postgres, Connection as _, PgConnection};
 // Workspace imports
 // Local imports
-use crate::connection::test_pool::TestPoolLock;
+use crate::connection::test_pool::{TestConnection, TestConnectionRef, TestTransaction};
+
+#[async_trait::async_trait]
+pub trait Acquire: Send {
+    async fn acquire(&mut self) -> ConnectionRef<'_>;
+    async fn begin(&mut self) -> sqlx::Result<Transaction<'_>>;
+}
 
 /// Connection holder unifies the type of underlying connection, which
 /// can be either pooled or direct.
-pub enum ConnectionHolder<'a> {
+pub enum Connection {
     Pooled(PoolConnection<Postgres>),
     Direct(PgConnection),
-    Transaction(Transaction<'a, Postgres>),
-    TestTransaction(TestPoolLock),
+    Test(TestConnection),
+}
+
+pub enum Transaction<'a> {
+    Real(sqlx::Transaction<'a, Postgres>),
+    Test(TestTransaction),
+}
+
+#[async_trait::async_trait]
+impl Acquire for Connection {
+    async fn acquire(&mut self) -> ConnectionRef<'_> {
```

Comment:
Wrapping multiple connections into non-commiting transactions in a single test will make those connections pretend to act on "separate databases" which will not work with our tests.

Reply:
Fetching a connection from a ConnectionPool may fail, while acquire() just dereferences the Connection (it always succeeds and in prod code is nonblocking). Making StorageProcessor wrap ConnectionPool rather than Connection will change the locations where processing may experience additional latency/starvation. With this pr StorageProcessor can wrap a Transaction or a Connection, while with your proposal it would wrap either Transaction or a ConnectionPool which IMO would look even weirder. Acquire is used only as a marker trait outside of dal crate. `sqlx::Acquire` is implemented not only for PgPool but also for PgConnection and Transaction, so the abstraction in this pr doesn't diverge from the sqlx interpretation in that aspect.